File: _inc/laravel/app/Http/Controllers/Individuals/CompetenciesController.php
<?php

namespace App\Http\Controllers\Individuals;

use App\Http\Controllers\Abstracts\Controller;

use App\Models\{Competencies, PerformanceType};
use Illuminate\Auth\Access\AuthorizationException;
use App\Config\Constants\{DatabaseConstants as DC, ViewsConstants as VW};
use Illuminate\Http\{RedirectResponse, Request};
use Illuminate\Support\Facades\Validator;
use Illuminate\Support\Facades\View as ViewFacade;

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
class CompetenciesController extends Controller
{
    private const PERM_MANAGE = 'Manage Competencies';
    private const PERM_CREATE = 'Create Competencies';
    private const PERM_EDIT   = 'Edit Competencies';
    private const PERM_DELETE = 'Delete Competencies';

    /** @return \Illuminate\View\View|RedirectResponse */
    public const IDX = 'index';
    public const CRT = 'create';
    public const STR = 'store';
    public const SHW = 'show';
    public const EDT = 'edit';
    public const UPD = 'update';
    public const DEL = 'destroy';

    public function index(Request $req)
    {
        $action = 'CompetenciesController@index';
        return $this->measureProfile($action, function () use ($req, $action) {
            if (($auth = self::authOrDeny($req, self::PERM_MANAGE, $action)) !== true) return $auth;

            $t = microtime(true);
            $competencies = Competencies::where(
                DC::COL_TABLE_CREATOR,
                $req->user()->creatorId()
            )->get();
            $this->logExecutionTime($t, $action . '::query', 'completed');

            $view = VW::CPT . '.index';
            if (!ViewFacade::exists($view)) abort(404, "View [$view] not found");

            return view($view, compact('competencies'));
        }, ['uri' => $req->getRequestUri(), 'ip' => $req->ip()]);
    }

    /** @return \Illuminate\View\View|RedirectResponse */
    public function create(Request $req)
    {
        $action = 'CompetenciesController@create';
        return $this->measureProfile($action, function () use ($req, $action) {
            if (($auth = self::authOrDeny($req, self::PERM_CREATE, $action)) !== true) return $auth;

            $t = microtime(true);
            $performanceTypes = PerformanceType::where(
                DC::COL_TABLE_CREATOR,
                $req->user()->creatorId()
            )->pluck('name', 'id')->prepend('Select Type', '');
            $this->logExecutionTime($t, $action . '::loadPerformanceTypes', 'completed');

            $view = VW::CPT . '.create';
            if (!ViewFacade::exists($view)) abort(404, "View [$view] not found");

            return view($view, compact('performanceTypes'));
        }, ['uri' => $req->getRequestUri()]);
    }

    /** @return RedirectResponse|null */
    public function store(Request $req)
    {
        $action = 'CompetenciesController@store';
        return $this->measureProfile($action, function () use ($req, $action) {
            if (($auth = self::authOrDeny($req, self::PERM_CREATE, $action)) !== true) return $auth;

            $t = microtime(true);
            $resp = self::validateInput($req, ['name' => 'required', 'type' => 'required']);
            $this->logExecutionTime($t, $action . '::validate', $resp ? 'failed' : 'completed');
            if ($resp) return $resp;

            try {
                $t = microtime(true);
                Competencies::create([
                    'name'       => $req->name,
                    'type'       => $req->type,
                    DC::COL_TABLE_CREATOR => $req->user()->creatorId(),
                ]);
                $this->logExecutionTime($t, $action . '::persist', 'completed');

                return redirect()->route(VW::CPT . '.index')
                    ->with('success', __('Competencies successfully created.'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($req, $e, $action);
            }
        }, ['uri' => $req->getRequestUri()]);
    }

    /** @return \Illuminate\View\View|RedirectResponse */
    public function show(Competencies $competency)
    {
        $action = 'CompetenciesController@show';
        // your original just redirected — keep it, but still measure
        return $this->measureProfile($action, function () {
            return redirect()->route(VW::CPT . '.index');
        });
    }

    /** @return \Illuminate\View\View|RedirectResponse */
    public function edit(Request $req, Competencies $competency)
    {
        $action = 'CompetenciesController@edit';
        return $this->measureProfile($action, function () use ($req, $competency, $action) {
            if (($auth = self::authOrDeny($req, self::PERM_EDIT, $action)) !== true) return $auth;

            $t = microtime(true);
            $ownerOk = ($competency->created_by === $req->user()->creatorId());
            $this->logExecutionTime($t, $action . '::authorizeOwner', $ownerOk ? 'ok' : 'denied');
            if (!$ownerOk) {
                return defaultPermissionDenial($req, new AuthorizationException(), $action);
            }

            $t = microtime(true);
            $performanceTypes = PerformanceType::where(
                DC::COL_TABLE_CREATOR,
                $req->user()->creatorId()
            )->pluck('name', 'id')->prepend('Select Type', '');
            $this->logExecutionTime($t, $action . '::loadPerformanceTypes', 'completed');

            $view = VW::CPT . '.edit';
            if (!ViewFacade::exists($view)) abort(404, "View [$view] not found");

            return view($view, compact('competency', 'performanceTypes'));
        }, ['competency_id' => $competency->id, 'uri' => $req->getRequestUri()]);
    }

    /** @return RedirectResponse|null */
    public function update(Request $req, Competencies $competency)
    {
        $action = 'CompetenciesController@update';
        return $this->measureProfile($action, function () use ($req, $competency, $action) {
            if (($auth = self::authOrDeny($req, self::PERM_EDIT, $action)) !== true) return $auth;

            $t = microtime(true);
            $ownerOk = ($competency->created_by === $req->user()->creatorId());
            $this->logExecutionTime($t, $action . '::authorizeOwner', $ownerOk ? 'ok' : 'denied');
            if (!$ownerOk) {
                return defaultPermissionDenial($req, new AuthorizationException(), $action);
            }

            $t = microtime(true);
            $resp = self::validateInput($req, ['name' => 'required', 'type' => 'required']);
            $this->logExecutionTime($t, $action . '::validate', $resp ? 'failed' : 'completed');
            if ($resp) return $resp;

            try {
                $t = microtime(true);
                $competency->update([
                    'name' => $req->name,
                    'type' => $req->type,
                ]);
                $this->logExecutionTime($t, $action . '::persist', 'completed');

                return redirect()->route(VW::CPT . '.index')
                    ->with('success', __('Competencies successfully updated.'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($req, $e, $action);
            }
        }, ['competency_id' => $competency->id, 'uri' => $req->getRequestUri()]);
    }

    /** @return RedirectResponse|null */
    public function destroy(Request $req, Competencies $competency)
    {
        $action = 'CompetenciesController@destroy';
        return $this->measureProfile($action, function () use ($req, $competency, $action) {
            if (($auth = self::authOrDeny($req, self::PERM_DELETE, $action)) !== true) return $auth;

            $t = microtime(true);
            $ownerOk = ($competency->created_by === $req->user()->creatorId());
            $this->logExecutionTime($t, $action . '::authorizeOwner', $ownerOk ? 'ok' : 'denied');
            if (!$ownerOk) {
                return defaultPermissionDenial($req, new AuthorizationException(), $action);
            }

            try {
                $t = microtime(true);
                $competency->delete();
                $this->logExecutionTime($t, $action . '::delete', 'completed');

                return redirect()->route(VW::CPT . '.index')
                    ->with('success', __('Competencies successfully deleted.'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($req, $e, $action);
            }
        }, ['competency_id' => $competency->id, 'uri' => $req->getRequestUri()]);
    }

    /**
     * Permission helper that returns true if authorized,
     * or a RedirectResponse from defaultPermissionDenial when not.
     */
    private static function authOrDeny(Request $req, string $perm, string $action): bool|RedirectResponse
    {
        if ($req->user()->can($perm)) return true;

        return defaultPermissionDenial(
            $req,
            new AuthorizationException(),
            $action
        );
    }

    private static function validateInput(Request $req, array $rules): ?RedirectResponse
    {
        $v = Validator::make($req->all(), $rules);
        return $v->fails()
            ? redirect()->back()->with('error', $v->getMessageBag()->first())
            : null;
    }
}

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/individuals/competencies_controller.py
import logging
from django.contrib import messages
from django.http import HttpRequest, HttpResponse, JsonResponse
from django.shortcuts import get_object_or_404, redirect, render
from ....Models.activity.performance_type import PerformanceType
from ....Models.individuals.competencies import Competencies
from .._traits.controller import Controller
from .._helpers.http import get_redirect_url
from django.core.exceptions import PermissionDenied
from ....configs.messages_templates import get_exception_class_message
logger = logging.getLogger(__name__)
class CompetenciesController(Controller):
  def index(self, request: HttpRequest) -> HttpResponse:
    try:
      return (render(request, 'competencies/index.html', {'competencies': Competencies.objects.filter(created_by=request.user.creator_id())})
              if request.user.has_perm('Manage Competencies')
              else (messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__)), redirect(get_redirect_url(request)))[1])  # Logic should be refactored
    except Exception as e:
      logger.exception(f"CompetenciesController.index failed: {e}")
      messages.error(request, "An error occurred while fetching competencies.")
      return redirect(get_redirect_url(request))
  def create(self, request: HttpRequest) -> HttpResponse:
    try:
      performance_qs = PerformanceType.objects.filter(created_by=request.user.creator_id())
      performance = {pt.id: pt.name for pt in performance_qs}
      return render(request, 'competencies/create.html', {'performance': performance})
    except Exception as e:
      logger.exception(f"CompetenciesController.create failed: {e}")
      messages.error(request, "An error occurred.")
      return redirect(get_redirect_url(request))
  def store(self, request: HttpRequest) -> HttpResponse:
    try:
      return (redirect('competencies_index') 
              if request.user.has_perm('Create Competencies') and request.POST.get('name') and request.POST.get('type') 
              and (Competencies.objects.create(name=request.POST.get('name'), type=request.POST.get('type'), created_by=request.user.creator_id()) or True)
              and messages.success(request, "Competencies successfully created.")
              else (messages.error(request, "Name and Type are required."), redirect(get_redirect_url(request)))[1]
             ) if request.user.has_perm('Create Competencies') else (messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__)), redirect(get_redirect_url(request)))[1]
    except Exception as e:
      logger.exception(f"CompetenciesController.store failed: {e}")
      messages.error(request, "An error occurred while creating competencies.")
      return redirect(get_redirect_url(request))
  def show(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      return redirect('competencies_index')
    except Exception as e:
      logger.exception(f"CompetenciesController.show failed: {e}")
      messages.error(request, "An error occurred.")
      return redirect('competencies_index')
  def edit(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      competency = get_object_or_404(Competencies, pk=id)
      performance_qs = PerformanceType.objects.filter(created_by=request.user.creator_id())
      performance = {pt.id: pt.name for pt in performance_qs}
      return render(request, 'competencies/edit.html', {'competencies': competency, 'performance': performance})
    except Exception as e:
      logger.exception(f"CompetenciesController.edit failed: {e}")
      messages.error(request, "An error occurred while editing competencies.")
      return redirect(get_redirect_url(request))
  def update(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      return (redirect('competencies_index')
              if request.user.has_perm('Edit Competencies') and request.POST.get('name') and request.POST.get('type')
              and (lambda c: (setattr(c, 'name', request.POST.get('name')), setattr(c, 'type', request.POST.get('type')), c.save(), messages.success(request, "Competencies successfully updated."), True)(get_object_or_404(Competencies, pk=id))) 
              else (messages.error(request, "Name and Type are required."), redirect(get_redirect_url(request)))[1]
             ) if request.user.has_perm('Edit Competencies') else (messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__)), redirect(get_redirect_url(request)))[1]
    except Exception as e:
      logger.exception(f"CompetenciesController.update failed: {e}")
      messages.error(request, "An error occurred while updating competencies.")
      return redirect(get_redirect_url(request))
  def destroy(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      return (get_object_or_404(Competencies, pk=id).delete(), messages.success(request, "Competencies successfully deleted."), redirect('competencies_index'))[2] if request.user.has_perm('Delete Competencies') else (messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__)), redirect(get_redirect_url(request)))[1]
    except Exception as e:
      logger.exception(f"CompetenciesController.destroy failed: {e}")
      messages.error(request, "An error occurred while deleting competencies.")
      return redirect(get_redirect_url(request))
